packages/core/content-manager/server/controllers/relations.js

```diff
@@ -1,64 +1,122 @@
+        return ctx.forbidden();
+      }
     }
 
-    const attribute = modelDef.attributes[targetField];
+    const attribute = sourceModel.attributes[targetField];
     if (!attribute || attribute.type !== 'relation') {
-      return ctx.badRequest('targetField.invalid');
+      return ctx.badRequest("This relational field doesn't exist");
     }
 
-    const target = strapi.getModel(attribute.target);
+    const targetedModel = strapi.getModel(attribute.target);
+
+    const modelConfig = component
+      ? await getService('components').findConfiguration(sourceModel)
+      : await getService('content-types').findConfiguration(sourceModel);
+    const mainField = prop(`metadatas.${targetField}.edit.mainField`, modelConfig) || 'id';
+
+    const fieldsToSelect = ['id', mainField];
+    if (hasDraftAndPublish(targetedModel)) {
+      fieldsToSelect.push(PUBLISHED_AT_ATTRIBUTE);
+    }
 
-    if (!target) {
-      return ctx.notFound('target.notFound');
+    // TODO: for RBAC reasons, find a way to exclude filters that should not be there
+    // i.e. all filters except locale for i18n
+    const queryParams = defaultsDeep(
+      {
+        orderBy: mainField,
+      },
+      {
+        ...transformParamsToQuery(targetedModel.uid, query),
+        select: fieldsToSelect, // cannot select other fields as the user may not have the permissions
+      }
+    );
+
+    if (!isEmpty(idsToOmit)) {
+      addWhereClause(queryParams, { id: { $notIn: idsToOmit } });
     }
 
-    if (idsToOmit && Array.isArray(idsToOmit)) {
-      query.filters = {
-        $and: [
-          {
-            id: {
-              $notIn: idsToOmit,
-            },
-          },
-        ].concat(query.filters || []),
-      };
+    // searching should be allowed only on mainField for permission reasons
+    if (_q) {
+      addWhereClause(queryParams, { [mainField]: { $containsi: _q } });
     }
 
-    const entityManager = getService('entity-manager');
+    if (entityId) {
+      const subQuery = strapi.db.queryBuilder(sourceModel.uid);
 
-    const entities = await entityManager.find(query, target.uid, []);
+      const alias = subQuery.getAlias();
 
-    if (!entities) {
-      return ctx.notFound();
-    }
+      const knexSubQuery = subQuery
+        .where({ id: entityId })
+        .join({ alias, targetField })
+        .select(`${alias}.id`)
+        .getKnexQuery();
 
-    const modelConfig = _component
-      ? await getService('components').findConfiguration(modelDef)
-      : await getService('content-types').findConfiguration(modelDef);
+      addWhereClause(queryParams, { id: { $notIn: knexSubQuery } });
+    }
 
-    const field = prop(`metadatas.${targetField}.edit.mainField`, modelConfig) || 'id';
-    const pickFields = [field, 'id', target.primaryKey, PUBLISHED_AT_ATTRIBUTE];
+    const results = await strapi.query(targetedModel.uid).findPage(queryParams);
 
-    ctx.body = entities.map(pick(pickFields));
+    ctx.body = {
```

Comment:
looks like you can just do ctx.body = results ?

Reply:
Almost as I think there is `pageCount` in the pagination that we don't want.
I prefer mapping to be in control of what is returned. But theoritically I could less map.
Wdyt?

Reply:
Just out of curiosity why not use the pageCount so we are consistent with other list APIs in the CM ?

Reply:
Oh, I didn't know the CM API has the `pageCount`, indeed. I'll put it then.